File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/application/usecases/GiftCardCreditingTest.java
package it.giannibombelli.wsc2026.giftcard.application.usecases;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.application.commands.CreditGiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;
import it.giannibombelli.wsc2026.giftcard.infrastructure.SqliteGiftCardRepository;
import it.giannibombelli.wsc2026.testsupport.AggregateFactory;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class GiftCardCreditingTest {
    private GiftCardRepository repository;
    private GiftCardCrediting crediting;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("giftcard");
        repository = new SqliteGiftCardRepository(dataSource);
        crediting = new GiftCardCrediting(repository);
    }

    @Nested
    class Construction {
        @Test
        void rejectNullRepository() {
            assertThatThrownBy(() -> new GiftCardCrediting(null))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class Execution {
        @Test
        void rejectNullCommand() {
            assertThatThrownBy(() -> crediting.invoke(null))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldPersistCredit() {
            GiftCard giftCard = AggregateFactory.getSavedGiftCard(repository);
            Money creditAmount = new Money(new BigDecimal("25.00"));

            crediting.invoke(new CreditGiftCard(giftCard.id(), creditAmount));

            GiftCard persisted = repository.findById(giftCard.id()).orElseThrow();
            assertThat(persisted.balance()).isEqualTo(creditAmount);
        }

        @Test
        void shouldFailIfCardDoesNotExist() {
            GiftCardId nonExisting = EntityId.generate(GiftCardId::new);
            Money amount = new Money(new BigDecimal("10.00"));

            assertThatThrownBy(() -> crediting.invoke(new CreditGiftCard(nonExisting, amount)))
                .isInstanceOf(IllegalStateException.class);
        }
    }
}

File: typescript/tests/giftcard/application/usecases/giftCardCrediting.test.ts
import { beforeAll, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import { CreditGiftCard, creditGiftCard } from '@/giftcard/application/commands/creditGiftCard.js';
import { GiftCardCrediting } from '@/giftcard/application/usecases/giftCardCrediting.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';
import { SqliteGiftCardRepository } from '@/giftcard/infrastructure/sqliteGiftCardRepository.js';
import { DatabaseSetup } from '../../../testsupport/databaseSetup.js';
import { GiftCardAggregateFactory } from '../../../testsupport/giftcard/aggregateFactory.js';

describe('GiftCardCrediting', () => {
  let repository: SqliteGiftCardRepository;
  let crediting: GiftCardCrediting;

  beforeAll(() => {
    const database = DatabaseSetup.initializeInMemoryDb('giftcard');
    repository = new SqliteGiftCardRepository(database);
    crediting = new GiftCardCrediting(repository);
  });

  describe('construction', () => {
    it('rejects null repository', () => {
      expect(() => new GiftCardCrediting(null as unknown as SqliteGiftCardRepository)).toThrow();
    });
  });

  describe('execution', () => {
    it('rejects null command', () => {
      expect(() => crediting.invoke(null as unknown as CreditGiftCard)).toThrow();
    });

    it('should persist credit', () => {
      const giftCard = GiftCardAggregateFactory.getSavedGiftCard(repository);
      const creditAmount = new Money(25);

      crediting.invoke(creditGiftCard(giftCard.id(), creditAmount));

      const persisted = repository.findById(giftCard.id());
      expect(persisted).not.toBeNull();
      expect(persisted!.balance()).toEqual(creditAmount);
    });

    it('should fail if card does not exist', () => {
      const nonExisting = generateId((value) => new GiftCardId(value));
      const amount = new Money(10);

      expect(() => crediting.invoke(creditGiftCard(nonExisting, amount))).toThrow();
    });
  });
});
